pkg/apis/config/defaults.go

```diff
@@ -97,21 +103,66 @@ type BrokerConfig struct {
 	Delivery           *eventingduckv1.DeliverySpec `json:"delivery,omitempty"`
 }
 
-// GetBrokerConfig returns a namespace specific Broker Configuration, and if
-// that doesn't exist, return a Cluster Default and if that doesn't exist
-// return an error.
-func (d *Defaults) GetBrokerConfig(ns string) (*BrokerConfig, error) {
+func (d *Defaults) GetBrokerConfig(ns string, brokerClassName string) (*BrokerConfig, error) {
 	if d == nil {
 		return nil, errors.New("Defaults are nil")
 	}
-	value, present := d.NamespaceDefaultsConfig[ns]
-	if present && value.BrokerConfig != nil {
-		return value.BrokerConfig, nil
+
+	// Early return if brokerClassName is provided and valid
+	if brokerClassName != "" {
+		return d.getBrokerConfigByClassName(ns, brokerClassName)
 	}
-	if d.ClusterDefault != nil && d.ClusterDefault.BrokerConfig != nil {
-		return d.ClusterDefault.BrokerConfig, nil
+
+	// Handling empty brokerClassName
+	return d.getBrokerConfigForEmptyClassName(ns)
+}
+
+func (d *Defaults) getBrokerConfigByClassName(ns string, brokerClassName string) (*BrokerConfig, error) {
+	// Check namespace specific configuration
+	if nsConfig, ok := d.NamespaceDefaultsConfig[ns]; ok && nsConfig != nil {
+		// check if the brokerClassName is the default broker class for this namespace
+		if nsConfig.DefaultBrokerClass == brokerClassName {
+			if nsConfig.BrokerConfig == nil {
+				// as no default broker class config is set for this namespace, should get the cluster default config
+				return d.getClusterDefaultBrokerConfig(brokerClassName)
+			}
+			return nsConfig.BrokerConfig, nil
+		}
+		if config, ok := nsConfig.BrokerClasses[brokerClassName]; ok && config != nil {
+			return config, nil
+		}
 	}
-	return nil, errors.New("Defaults for Broker Configurations have not been set up.")
+
+	// Check cluster default configuration
+	return d.getClusterDefaultBrokerConfig(brokerClassName)
+}
+
+func (d *Defaults) getBrokerConfigForEmptyClassName(ns string) (*BrokerConfig, error) {
+	// Check if namespace has a default broker class
+	if nsConfig, ok := d.NamespaceDefaultsConfig[ns]; ok && nsConfig != nil {
+		if nsConfig.DefaultBrokerClass != "" {
+			return d.getBrokerConfigByClassName(ns, nsConfig.DefaultBrokerClass)
+		}
+	}
+
+	// Fallback to cluster default configuration
+	return d.getClusterDefaultBrokerConfig("")
+}
+
+func (d *Defaults) getClusterDefaultBrokerConfig(brokerClassName string) (*BrokerConfig, error) {
+	if d.ClusterDefaultConfig == nil || d.ClusterDefaultConfig.BrokerConfig == nil {
+		return nil, errors.New("Defaults for Broker Configurations have not been set up.")
+	}
+
+	if brokerClassName == "" {
+		return d.ClusterDefaultConfig.BrokerConfig, nil
+	}
+
+	if config, ok := d.ClusterDefaultConfig.BrokerClasses[brokerClassName]; ok && config != nil {
```

Comment:
shouldn't we check before, if `d.ClusterDefaultConfig.BrokerClass == brokerClassName` and if this is the case return `d.ClusterDefaultConfig.BrokerConfig`, as you wrote in https://docs.google.com/document/d/1RKij-DYPmcbCTHF26hkXdrtWHqrksA-Fo5qYLRP7xVA/edit
```
# The brokerClasses contain the configmap, delivery spec for other brokerClasses.
# The one specified above is NOT included.
# Reason: to maintain compatibility, we are not removing those old fields. And it is not 
# necessary to add it again here under brokerClasses.
```

Otherwise `d.ClusterDefaultConfig.BrokerConfig` does not take preference over `d.ClusterDefaultConfig.BrokerClasses[]` (what I meant in https://github.com/knative/eventing/pull/7631#discussion_r1475660596)

Reply:
I think the logic here is correct, prob I didn't explain clearly in the Doc, sorry about that! @creydr 

if `d.ClusterDefaultConfig.BrokerClass == brokerClassName`, then we will directly use `d.ClusterDefaultConfig.BrokerConfig`. And this brokerClass shouldn't appear in `d.ClusterDefaultConfig.BrokerClasses`.

When I was referring to:
> The one specified above is NOT included.

It means the brokerConfig for `d.ClusterDefaultConfig.BrokerClass` will not be included in the `d.ClusterDefaultConfig.BrokerClasses`